modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -470,34 +477,53 @@ def map_func(left, right=right, kwargs=kwargs):
                     keep_partitioning=False,
                 )
             )
-            is_reset_index = True
-            if left_on and right_on:
-                left_on = left_on if is_list_like(left_on) else [left_on]
-                right_on = right_on if is_list_like(right_on) else [right_on]
-                is_reset_index = (
-                    False
-                    if any(o in new_self.index.names for o in left_on)
-                    and any(o in right.index.names for o in right_on)
-                    else True
-                )
-                if sort:
+
+            # Here we want to understand whether we're joining on a column or on an index level.
+            # It's cool if indexes are already materialized so we can easily check that, if not
+            # it's fine too, we can also decide that by columns, which tend to be already
+            # materialized quite often compared to the indexes.
+            keep_index = False
+            if self._modin_frame._index_cache is not None:
+                if left_on is not None and right_on is not None:
+                    keep_index = any(
+                        o in self.index.names
+                        and o in right_on
+                        and o in right.index.names
```

Comment:
```suggestion
                        o in self.index.names
                        and (o in right_on
                        or o in right.index.names)
```
Should we check this way?

Reply:
No, it's mandatory for both keys to be in their `on` argument and to be in an actual index level
```python
# 'idx_key2' exists in both indices but is specified only in a 'left_on'
>>> df1i.index.names
FrozenList(['idx_key1', 'idx_key2'])
>>> df2i.index.names
FrozenList(['idx_key2', 'idx_key3'])
>>> df1i.merge(df2i, left_on=["idx_key2"], right_on=["idx_key3"])
   idx_key3  data_col1  col_key1_x  col_key2_x  idx_key1  data_col2  col_key1_y  col_key2_y
0         4          2           4           4         1          4           3           3
1         5          3           5           5         2          3           4           4
2         6          4           6           6         3          2           5           5
# 'idx_key1' exists in both frame, in the left it's an index level and in the right it's a column name
>>> df1i.index.names
FrozenList(['idx_key1', 'idx_key2'])
>>> df2i.columns
Index(['idx_key1', 'data_col2', 'col_key1', 'col_key2'], dtype='object')
>>> df1i.merge(df2i, left_on=["idx_key1"], right_on=["idx_key1"])
   idx_key1  idx_key3  data_col1  col_key1_x  col_key2_x  data_col2  col_key1_y  col_key2_y
0         1         3         10           3           3          4           3           3
1         2         4          2           4           4          3           4           4
2         3         5          3           5           5          2           5           5
3         4         6          4           6           6         10           6           6
# 'idx_key2' exists in both frames as an index level and is specified in both 'left_on' and 'right_on'
>>> df1i.merge(df2i, left_on=["idx_key2"], right_on=["idx_key2"])
          idx_key3  data_col1  col_key1_x  col_key2_x  idx_key1  data_col2  col_key1_y  col_key2_y
idx_key2
2                3         10           3           3         1          4           3           3
3                4          2           4           4         2          3           4           4
4                5          3           5           5         3          2           5           5
5                6          4           6           6         4         10           6           6
```